### PR TITLE
Installation and Testing Improvements

### PR DESCRIPTION
**Pull request by TheForgotten69**
## Issues Found
1. Model Registration Mismatches
   - Architecture name mismatch in checkpoint: expected 'IPEC-decoder_dctform-rgb-share-size4'
   - Required adding correct registration decorator in JDEC.py:
     ```python
     @register('IPEC-decoder_dctform-rgb-share-size4')  # Changed from 'jdec'
     class JDEC(nn.Module):
     ```
   - Secondary mismatch with 'swinv2_rgb_nomore_deepir_rgbver' encoder model
   - Quick fix implemented by using existing 'swinv2_group_embedded' architecture which shares the same underlying structure

2. Environment Setup Issues
   - Missing directory structure for temporary files
   - Added missing `bin` directory creation for JPEG processing through gitkeep

## Testing Performed

Tested on Ubuntu 20.04 environment
Successfully processed test images

### CODE REVIEW

**Comment by Arcitec** [comment]
@TheForgotten69 Thank you so much for working on this and discovering these errors in the repo!

I downloaded the fixed checkpoint file (thanks @WooKyoungHan !) and still couldn't run it, because it said some error related to `IPEC-decoder_dctform-rgb-share-size4`. But with these patches, it finally works!

I can confirm my list of available models is the same as yours:

`Available models: dict_keys(['mlp', 'mlp_1dconv', 'swinv2_group_embedded', 'IPEC-decoder_dctform-rgb-share-size4'])`

@WooKyoungHan Do you have any ideas why `swinv2_rgb_nomore_deepir_rgbver` does not exist for us but `swinv2_group_embedded` exists? I think we might be using the wrong model. I think `nomore` patches is a big part of JDEC's architecture, isn't it? (Edit: Yes [RGB No More](https://github.com/JeongsooP/RGB-no-more) is very important. We need to fix this missing model, I think...)

But at least, these patches allowed me to do some first testing with JDEC and I am really impressed. It totally removes the artifacts. I will be watching this pull request with big interest so we can figure out the final problems here! :)

If you have some local code that is not merged and fixes these problems, please be sure to upload it to GitHub in case this Pull Request is not necessary with your own local code. :)

**Comment by Arcitec** [comment]
I just did another quick check and see that `swinv2_group_embedded` comes from these code locations:

```
JDEC/models/swinirv2.py
901:@register('swinv2_group_embedded')

JDEC/configs/train_JDEC.yaml
37:      name: swinv2_group_embedded
```

That means the JDEC project bundles this model in its own source code, and it's named `swinv2_group_embedded`.

So it seems like that may be the correct model name after all? So we don't need to find `swinv2_rgb_nomore_deepir_rgbver`? Please investigate @WooKyoungHan  :)

We really need to be 1000000% sure that the repository code is fully updated to match your local code, since all these errors seems like the current repo is not the fixed code that you have locally. ;)

**Comment by WooKyoungHan** [comment]
This is a problem that occurred when I uploaded it without updating it. This is a case in which an old version of the register name was applied. It can be solved simply as above, but it seems cleaner from previous experience to solve it at the register declaration stage. I will organize it well and upload it right away. I want to express my sincere gratitude to all of you who are interested in my thesis.

**Comment by Arcitec** [comment]
@WooKyoungHan Thank you, I am happy to hear this. And I agree, it's better to solve this naming problem at the `@register` and `name:` config, because the `swinv2_rgb_nomore_deepir_rgbver` name is more accurate to be using.

The `@register('IPEC-decoder_dctform-rgb-share-size4')` also needs a fix somewhere (I am guessing this change needs to be included since it seems like that's the name inside the checkpoint file).

I have tried a few quick tests of the model at q30 and q85 by the way. It's superb at removing JPEG artifacts, but I am a bit worried that it also removes a lot of high frequency details (like skin texture) even at q85 JPEG input, so I am wondering if that's normal or if something can be done to keep more high frequency details?

Edit: I might be wrong. Maybe I made a mistake and looked at the wrong images. I am gonna do some more testing! :)

**Comment by WooKyoungHan** [comment]
Perhaps my method of solving the part is more of a temporary method than a fundamental problem. In this regard, the fundamental way to solve the problem is re-training with 'jdec,' which is the most appropriate method (I apologize for this and will consider re-learning within a few weeks). 

--------
Thanks for bringing up this issue.

These days, I am thinking a lot about the high-frequency details problem. This is because increasing the PSNR (reducing errors in per-pixel units) is not equivalent to increasing the high-frequency details (although it is highly correlated). Further, jdec is an INR-based network, which is known to be vulnerable to this problem (spectral bias). Although JDEC mitigated this using the method CCF, the unresolved spectral biases will cause unsatisfactory results.

So, how do we solve this problem? The fastest and most effective way would be to add the GAN method. Many studies have reported that the same method as the GAN works well in this respect. Also, GAN discriminators are known to 'like' high-frequencies (but in fact, discriminators are also difficult to judge high-frequency as fake). Although this may not be good in terms of restoring real images, it seems to be a better way to observe more 'fine' images in our eyes. This is going to be a question of which metric is better for us.


I'm thinking about how to restore high frequency through GAN-free methods. In most cases, the fact is that if we make the network huge, we'll have fewer errors in the per-pixel units and better details, so it's good to express high-frequency. But it's hard to say that this is methodologically efficient. 

The kernel method is the one I thought about the most. Although it is necessary to consider whether the kernel method is effective in jpeg, which is an ill-posed problem, I think it would be fun to think about generating high-frequency random components and cut them well with the kernel to make details.

**Comment by Arcitec** [comment]
> Perhaps my method of solving the part is more of a temporary method than a fundamental problem. In this regard, the fundamental way to solve the problem is re-training with 'jdec,' which is the most appropriate method (I apologize for this and will consider re-learning within a few weeks).

Ahh. Is there any quality-problem with the current checkpoint file, or is it only "wrong" "layer name"?

If you only want to change the layer name to something that's "nicer sounding" ("jdec"), maybe you can easily rename the layer inside the checkpoint file without retraining it?

Edit: It [might be possible](https://discuss.pytorch.org/t/how-to-change-the-name-of-the-weights-to-a-new-name-when-saving-the-model/32166/12) to "load model, rename field, save model", but they warn that it may have "some side effects".

**Comment by Arcitec** [comment]
Well, I see that the paper says:

> Implementation Detail
> 
> We use 112 × 112 patches as inputs to our network. This size is chosen because it is the
> least common multiple of the minimum unit size of color
> JPEG (16 × 16) and the window size of our Swin architecture [27] (7 × 7). The network is trained for 1000 epochs
> with batch size 16. We optimize our network by Adam [20].
> The learning rate is initialized as 1e-4 and decayed by factor
> 0.5 at [200, 400, 600, 800].

So it seems like it can use different input sizes. But that the JPEG files dimensions must be multiples of 112?

I hope you can clarify whether this network can:

- Use arbitrary input dimensions which are not multiple of 112, such as 707 x 1231? In other words, real-world JPEGs.
- Can use large input dimensions like 6k x 6k?

**Comment by TheForgotten69** [comment]
@WooKyoungHan What do you think of this PR? It's still worth to be merge imo as the current code won't work as is.
